### PR TITLE
[4/n][torchtune integration] support lazy load model during inference

### PR DESCRIPTION
**Pull request by SLR722**
## What does this PR do?
In this PR, we refactor the meta reference inference logic to support 
- load the model during registering model instead of during spinning up server
- support inference finetuned model checkpoint on top of native llama model

## Why need these changes
To solve the existing pain points that 
- user cannot lazy load the model and hot switch the inference checkpoint after spinning up the server 
  - this blocks us doing inference and eval on the same sever for a finetuned checkpoint after post training
 - user cannot do inference on a finetuned checkpoint on top of native llama models

## Expect user experience change
- The inference model won't be loaded when spinning up server. Instead, it will be loaded during register model. If user add the model as models resource in run.yaml, it will be registered and loaded automatically when starting server. There is  an optional flag 'skip_initialize' in model metadata to skip model loading during registration. 
- There is an optional flag 'llama_model' in model metadata to identify the base model of the Model class for validation and initialize model arch. model identifier no longer needs to be a native llama model
- the default inference model name updates from 'meta-llama/Llama-3.2-3B-Instruct' to 'Llama3.2-3B-Instruct' 
  - It aligns with the checkpoint folder name after running 'llama model download'
  - It aligns with the descriptor name defined in llama-models SKU list https://github.com/meta-llama/llama-models/blob/bf5b0c4fe74e3b51ed5904ab65e3f671b194d2a9/models/datatypes.py#L95

## test
run python llama_stack/scripts/distro_codegen.py

**run unit test**
- torchrun $CONDA_PREFIX/bin/pytest -v -s -k "meta_reference" --inference-model="Llama3.1-8B-Instruct" ./llama_stack/providers/tests/inference/test_text_inference.py
- torchrun $CONDA_PREFIX/bin/pytest -v -s -k "meta_reference" --inference-model="Llama3.1-8B-Instruct" ./llama_stack/providers/tests/inference/test_model_registration.py

**test post training experience**
on server side run: llama stack run llama_stack/templates/experimental-post-training/run.yaml
server is spinning up without model loaded

Screenshot 2024-12-17 at 1 24 50 PM

on client side, run:  llama-stack-client --endpoint http://devgpu018.nha2.facebook.com:5000 models register Llama3.2-3B-Instruct
register model successfully and the model is loaded 
Screenshot 2024-12-17 at 1 26 30 PM

Screenshot 2024-12-17 at 1 26 09 PM

if add "skip_initialize" in metadata, model is registered but isn't loaded

on client side, run:  llama-stack-client --endpoint http://devgpu018.nha2.facebook.com:5000 inference chat-completion --message "hello, what model are you?"

Inference the model succesfully
Screenshot 2024-12-17 at 1 27 33 PM

**test inference experience**
run: llama stack run llama_stack/templates/meta-reference-gpu/run.yaml
model is loaded since the model is in resouce list in run.yaml 
Screenshot 2024-12-17 at 1 30 19 PM

on client side, run: llama-stack-client --endpoint http://devgpu018.nha2.facebook.com:5000 inference chat-completion --message "hello, what model are you?"
inference successfully 
Screenshot 2024-12-17 at 1 31 08 PM

## inference on a finetuned model
**register a finetuned model that finetuned by post training api (torchtune)**
- the model is registered and loaded successfully 
- the model is shown up in the model list 
Screenshot 2024-12-18 at 3 56 33 PM

**run inference**

Screenshot 2024-12-18 at 3 57 59 PM

### CODE REVIEW

File: llama_stack/templates/meta-reference-gpu/meta_reference.py [hunk 1]
```diff
@@ -112,15 +112,15 @@ def get_distribution_template() -> DistributionTemplate:
                 "Port for the Llama Stack distribution server",
             ),
             "INFERENCE_MODEL": (
-                "meta-llama/Llama-3.2-3B-Instruct",
+                "Llama3.2-3B-Instruct",
```
Comment (ashwinb) [inline llama_stack/templates/meta-reference-gpu/meta_reference.py:115]: why did you change these?

Comment (SLR722) [reply llama_stack/templates/meta-reference-gpu/meta_reference.py:115]: We want to use the model_id name as checkpoint path name to read checkpoint from. Changing the name make 
- It aligns with the checkpoint folder name after running 'llama model download'
- It aligns with the descriptor name defined in llama-models SKU list https://github.com/meta-llama/llama-models/blob/bf5b0c4fe74e3b51ed5904ab65e3f671b194d2a9/models/datatypes.py#L95

The original name "meta-llama/Llama-3.2-3B-Instruct" is the hugging face repo name which I feel is a little bit weird here as 'llama model download' is default to download model from meta

File: llama_stack/providers/inline/inference/meta_reference/inference.py [hunk 2]
```diff
@@ -71,50 +70,66 @@ class MetaReferenceInferenceImpl(
...
-            ],
-        )
-        self.model = model
-        # verify that the checkpoint actually is for this model lol
+        self.model_id = None
+        self.llama_model = None
 
-    async def initialize(self) -> None:
-        log.info(f"Loading model `{self.model.descriptor()}`")
+    async def initialize(self, model_id, llama_model) -> None:
```
Comment (ashwinb) [inline llama_stack/providers/inline/inference/meta_reference/inference.py:76]: oh `initialize()` can never get an argument which is not specified in the config. the API is always:

```
# llama stack infra create a config for a provider 
# calls get_provider_impl(config)
# that method constructs the class 
# and initialize()s it
```

**Review by ashwinb — CHANGES_REQUESTED** [review]
adding an argument to `initialize()` is a no-go; you can refactor the methods -- call it `load_model()`

File: llama_stack/providers/inline/inference/meta_reference/__init__.py [hunk 3]
```diff
@@ -16,5 +16,5 @@ async def get_provider_impl(
     from .inference import MetaReferenceInferenceImpl
 
     impl = MetaReferenceInferenceImpl(config)
-    await impl.initialize()
```
Comment (ashwinb) [inline llama_stack/providers/inline/inference/meta_reference/__init__.py:19]: no! why was this removed?! you must keep it. if there is nothing to initialize() just add a `pass`? but keep the initializer there

**Review by ashwinb — APPROVED** [review]
thank you for the iteration <3
